### PR TITLE
Remove some deprecated stdlib files

### PR DESCRIPTION
**Pull request by Villetaneuse**
Removed 4 deprecated files in the stdlib (3697 lines).
- `Numbers/Int31/` is deprecated since 8.10.
- `Sorting/Heap.v` is deprecated since [this commit](https://github.com/coq/coq/commit/f698148f6aee01a207ce5ddd4bebf19da2108bff) of 2009

- [ ] Added **changelog**.
- [ ] Added / updated **documentation**.

### CODE REVIEW

**Comment by olaure01** [comment]
There is a number of files in the standard library which contain a comment mentioning them as deprecated. However this is informal and does not generate any warning to users.
An [explicit deprecation phase](https://coq.github.io/doc/master/refman/using/libraries/writing.html#deprecating-library-objects-or-tactics) is probably better before removal.
Deprecation warnings are [introduced in Coq 8.18](https://github.com/coq/coq/pull/17734) for the `Numbers/Cyclic/Int31/` files but there is no such warning in `Sorting\Heap.v`.

**Comment by andres-erbsen** [comment]
FWIW I am in favor of removal without deprecation unless the CI shows widespread breakage or defensible use of the removed definitions in inter-package interfaces. Developing the standard library is hard enough already, and downstream developments can just vendor files like this if they want theory no longer present in stdlib.

**Comment by andres-erbsen** [comment]
`dune build @stdlib-html` is failing, this can probably be fixed [here](https://github.com/coq/coq/blob/master/doc/stdlib/index-list.html.template#L275-L277).

**Comment by JasonGross** [comment]
Another option for compat, btw, is to move the files' contents to `Coq/Compat/Coq*.v` modules.  We should probably then make "unknown -compat version" a warning rather than an error.  Assuming `Require` always prefers absolute names, users wanting compat can then create an empty `Coq/*/*.v` files in their project, throw `-compat...` into their flags, and keep using the files for another couple versions.

**Comment by JasonGross** [comment]
(To be clear though, I'm fine with removing unused files without an explicit deprecation phase.  My policy on compat is "it should ideally almost always be possible to automate backwards compatible fixes to downstream projects", and "vendor the files locally" is a fine solution as far as I'm concerned)

**Comment by Villetaneuse** [comment]
I'll keep this frozen then, waiting further instructions.
Is would be nice to have a `#[deprecated]` flag for whole files.

**Comment by olaure01** [comment]
> Is would be nice to have a `#[deprecated]` flag for whole files.

https://github.com/coq/coq/issues/8032

**Comment by Villetaneuse** [comment]
So, what do we make of this?
- Should I remove this PR and start a new one with deprecation warnings for all the elements of `Heap.v` or
- Try to fix the broken tests and see if it passes CI to remove the files?

**Comment by proux01** [comment]
In any case, you can keep this PR.
Since Int31 has already been deprecated properly in #17734 that is included in 8.18, you can simply remove the files in this PR that targets 8.19.
For heap.v, we are unfortunately still lacking a solution to deprecate a full module so the best solution is to deprecate each element in the file. It is small enough that this sould be reasonnable (particularly considering that we can now deprecate definitions (with the same syntax as notations: `#[deprecated(since="8.3")] Definition ...`)). The file could then be removed in 8.20 (in a further PR in six months in the future).

**Comment by Villetaneuse** [comment]
That's embarrassing... I think I did something very wrong with rebase. Very sorry about it. Can I fix that or should I close and reopen? Again, very sorry.

**Comment by SkySkimmer** [comment]
You can `git reset --hard` to the commit before the rebase (I guess 66f1d25 ?) and try to rebase again.

**Comment by jfehrle** [comment]
> That's embarrassing... I think I did something very wrong with rebase. Very sorry about it. Can I fix that or should I close and reopen? Again, very sorry.

No worries.  Everyone makes mistakes from time to time.  This one is readily apparent and easily fixed. No need for a new PR.

**Comment by Villetaneuse** [comment]
For what it's worth, I wrote down a small shell script to add deprecation warnings:
```sh
#!/bin/sh
usage()
{
	cat 1>&2 <<EOF
Usage: $0 FILE VERSION NOTE

	Write deprecation attributes in a Coq source file for all commands that
	support it.
	Example: $0 Heap.v 8.3 "Use mergesort.v"
EOF
}

case $# in
	3) ;;
	*) usage; exit 1
esac

file=$1
vers=$2
note=$3

if ! [ -f "$file" ]; then
	printf '%s: %s: No such file\n' "$0" "$file" 1>&2
	exit 1
fi

deprable="Theorem|Lemma|Fact|Corollary|Proposition|Property|\
Definition|Example|Fixpoint"

annot="#[deprecated(since=\"$vers\", note=\"$note\")]"

tmp="$file".depr

sed -E 's/(^[[:blank:]]*)(('"$deprable"').*$)/\1'"$annot"'\n\1\2/' \
	"$file" >"$tmp"

mv "$tmp" "$file"
```
If you want, I can add it in `dev/tools` (but then I guess there is a high probability that the next person to need it won't know it exists...).
It does not handle specially commands in comments (so if there are, there will be useless deprecated warning inside the comments too).

Also, the choice of what is "deprecatable" and what is not is not obvious to me. For instance I don't see why an inductive type which comes with 4 automatically generated lemmas cannot be deprecated. But it's probably a very naive comment.

**Comment by andres-erbsen** [comment]
I think it would be nice to have a deprecation script in the repo. I have been using `sed 's/^\(\s*\)\(\(Local\s+\|Global\s+\|Program\s+\|Canonical\+\)*\(Lemma\|Theorem\|Corollary\|Definition\|Fixpoint\|Example\|Instance\|Axiom\|Parameter\|Notation\|Coercion\)\)/\1#[deprecated(note="CHANGEME", since="NOW")]\n\1\2/g'`

**Comment by Villetaneuse** [comment]
> I think it would be nice to have a deprecation script in the repo. I have been using `sed 's/^\(\s*\)\(\(Local\s+\|Global\s+\|Program\s+\|Canonical\+\)*\(Lemma\|Theorem\|Corollary\|Definition\|Fixpoint\|Example\|Instance\|Axiom\|Parameter\|Notation\|Coercion\)\)/\1#[deprecated(note="CHANGEME", since="NOW")]\n\1\2/g'`

Alright, I'm modifying it with your additions (Instance, Axiom, etc) and I will add it to the repo.

**Comment by JasonGross** [comment]
> Also, the choice of what is "deprecatable" and what is not is not obvious to me. For instance I don't see why an inductive type which comes with 4 automatically generated lemmas cannot be deprecated. But it's probably a very naive comment.

Probably it's just that nobody has gotten around to writing the infrastructure to allow deprecating inductives, and the abstractions are not clean enough that deprecation support happens automatically.

**Comment by Villetaneuse** [comment]
I've added my small 'deprecate_file' script in dev/tools. I think I'm done here, unless someone has more suggestions. (I think the changelog entries come later ?)
The stdlib doc looks up to date.

File: dev/tools/deprecate_file [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#!/bin/sh
```
Comment (proux01) [inline dev/tools/deprecate_file:1]: Maybe the file can be named `deprecate_file.sh` like others in the same directory?

File: theories/Sorting/Heap.v [hunk 2]
```diff
@@ -88,13 +95,14 @@ Section defs.
...
   Lemma is_heap_rect :
     forall P:Tree -> Type,
       P Tree_Leaf ->
       (forall (a:A) (T1 T2:Tree),
-	leA_Tree a T1 ->
-	leA_Tree a T2 ->
-	is_heap T1 -> P T1 -> is_heap T2 -> P T2 -> P (Tree_Node a T1 T2)) ->
+        leA_Tree a T1 ->
+        leA_Tree a T2 ->
+        is_heap T1 -> P T1 -> is_heap T2 -> P T2 -> P (Tree_Node a T1 T2)) ->
```
Comment (proux01) [inline theories/Sorting/Heap.v:105]: This change seems unrelated to the current PR? No big deal but it clutters the changelog, so if it can be avoided, all the better.

File: theories/Sorting/Heap.v [hunk 3]
```diff
@@ -105,13 +113,14 @@ Section defs.
...
   Lemma is_heap_rec :
     forall P:Tree -> Set,
       P Tree_Leaf ->
       (forall (a:A) (T1 T2:Tree),
-	leA_Tree a T1 ->
-	leA_Tree a T2 ->
-	is_heap T1 -> P T1 -> is_heap T2 -> P T2 -> P (Tree_Node a T1 T2)) ->
+        leA_Tree a T1 ->
+        leA_Tree a T2 ->
+        is_heap T1 -> P T1 -> is_heap T2 -> P T2 -> P (Tree_Node a T1 T2)) ->
```
Comment (proux01) [inline theories/Sorting/Heap.v:123]: same

File: theories/Sorting/Heap.v [hunk 4]
```diff
@@ -139,13 +149,14 @@ Section defs.
       (forall a, HdRel leA a l1 -> HdRel leA a l2 -> HdRel leA a l) ->
       merge_lem l1 l2.
   Import Morphisms.
-  
+
```
Comment (proux01) [inline theories/Sorting/Heap.v:152]: Unrelated change (and since the plan is to remove the file, probably not worth it)

**Review by proux01 — APPROVED** [review]
If you want you can still deprecate the inductives by renaming them to something like `tree_deprecated` then adding a notation
```Coq
#[deprecated(since="8.3", note="Use mergesort.v")]
Notation tree := tree_deprecated.
```

File: theories/Sorting/Heap.v (hunk 2 above)
Comment (Villetaneuse) [reply theories/Sorting/Heap.v:105]: No sure if I can, this was done automatically by a hook when I committed.

File: theories/Sorting/Heap.v (hunk 3 above)
Comment (Villetaneuse) [reply theories/Sorting/Heap.v:123]: Same here.

File: theories/Sorting/Heap.v (hunk 4 above)
Comment (Villetaneuse) [reply theories/Sorting/Heap.v:152]: Same here.

File: dev/tools/deprecate_file (hunk 1 above)
Comment (Villetaneuse) [reply dev/tools/deprecate_file]: done (notice the tabs changed to spaces for indentation, this is done by git-fix-ws-tree, although I guess tabs in sh scripts are ok?).

File: theories/Sorting/Heap.v (hunk 2 above)
Comment (proux01) [reply theories/Sorting/Heap.v:105]: Nevermind, my point was just that if you can avoid that in future contributions, all the better since avoiding unrelated changes esaes the review process.

Comment (Villetaneuse) [reply theories/Sorting/Heap.v:105]: Sorry, I will look into it (check if this can be disabled somehow).

File: dev/tools/deprecate_file (hunk 1 above)
Comment (proux01) [reply dev/tools/deprecate_file]: I hope so

**Comment by Villetaneuse** [comment]
> If you want you can still deprecate the inductives by renaming them to something like `tree_deprecated` then adding a notation
> 
> ```coq
> #[deprecated(since="8.3", note="Use mergesort.v")]
> Notation tree := tree_deprecated.
> ```

Do you think I should do it, then? Or for now it's not worth the trouble?

File: theories/Sorting/Heap.v (hunk 2 above)
Comment (SkySkimmer) [reply theories/Sorting/Heap.v:105]: The hook is supposed to change tabs to spaces only for lines which were modified.

File: dev/tools/deprecate_file (hunk 1 above)
Comment (SkySkimmer) [reply dev/tools/deprecate_file]: Absence of tab checking is controlled by .gitattributes which is extension based, so an extension-less file avoids it.

**Comment by proux01** [comment]
So, apparently, you need to write overlays for
* coqprime
* fiat_parsers
* metacoq

to handle the removal of Int31 (pinging @JasonGross for the first two).

Let me know if you need any help.

**Comment by Villetaneuse** [comment]
CoqPrime should be ok... Is there a way to launch CI only for coqprime?

**Comment by proux01** [comment]
Unfortunately no

**Comment by Villetaneuse** [comment]
@JasonGross I'm a bit lost with `fiat_parsers`. There are many files importing `Int31`, mostly compatibility files.
```text
$ findgrep v 'nt31'
./etc/coq-scripts/compatibility/Coq__8_12__Compat.v:21:Require Coq.Numbers.Cyclic.Int31.Int31.
./etc/coq-scripts/compatibility/Coq__8_12__Compat.v:27:Numeral Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
./etc/coq-scripts/compatibility/Coq__8_13__Compat.v:21:Require Coq.Numbers.Cyclic.Int31.Int31.
./etc/coq-scripts/compatibility/Coq__8_13__Compat.v:26:Numeral Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
./etc/coq-scripts/compatibility/Coq__8_14__Compat.v:21:Require Coq.Numbers.Cyclic.Int31.Int31.
./etc/coq-scripts/compatibility/Coq__8_14__Compat.v:25:Number Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
./etc/coq-scripts/compatibility/Coq__8_15__Compat.v:21:Require Coq.Numbers.Cyclic.Int31.Int31.
./etc/coq-scripts/compatibility/Coq__8_15__Compat.v:25:Number Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
./etc/coq-scripts/compatibility/Coq__8_16__Compat.v:21:Require Coq.Numbers.Cyclic.Int31.Int31.
./etc/coq-scripts/compatibility/Coq__8_16__Compat.v:25:Number Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
./etc/coq-scripts/compatibility/Coq__8_17__Compat.v:21:Require Coq.Numbers.Cyclic.Int31.Int31.
./etc/coq-scripts/compatibility/Coq__8_17__Compat.v:25:Number Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
./etc/coq-scripts/compatibility/Coq__8_18__Compat.v:21:Require Coq.Numbers.Cyclic.Int31.Int31.
./etc/coq-scripts/compatibility/Coq__8_18__Compat.v:25:Number Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
./etc/coq-scripts/compatibility/Coq__8_19__Compat.v:22:Require Coq.Numbers.Cyclic.Int31.Int31.
./etc/coq-scripts/compatibility/Coq__8_19__Compat.v:26:Number Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
./etc/coq-scripts/compatibility/Coq__master__Compat.v:21:Require Coq.Numbers.Cyclic.Int31.Int31.
./etc/coq-scripts/compatibility/Coq__master__Compat.v:25:Number Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
./etc/coq-scripts/compatibility/fragments/Grab88v811.v:19:Require Coq.Numbers.Cyclic.Int31.Int31.
./etc/coq-scripts/compatibility/fragments/Grab88v811.v:25:Numeral Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
./etc/coq-scripts/compatibility/fragments/Grab88v813.v:19:Require Coq.Numbers.Cyclic.Int31.Int31.
./etc/coq-scripts/compatibility/fragments/Grab88v813.v:24:Numeral Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
./etc/coq-scripts/compatibility/fragments/Grab88v814.v:19:Require Coq.Numbers.Cyclic.Int31.Int31.
./etc/coq-scripts/compatibility/fragments/Grab88v814.v:23:Number Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
./etc/coq-scripts/compatibility/fragments/Grab88v816.v:19:Require Coq.Numbers.Cyclic.Int31.Int31.
./etc/coq-scripts/compatibility/fragments/Grab88v816.v:23:Number Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
./src/Common/Coq__8_4__8_5__Compat.v:22:Require Coq.Numbers.Cyclic.Int31.Int31.
./src/Common/Coq__8_4__8_5__Compat.v:26:Number Notation Int31.int31 Int31.phi_inv_nonneg Int31.phi : int31_scope.
```
I don't really know where they are used (a quick grep gave me way too many files). And Uint63 exists (with this name) only since 2021. I feel that I won't be able to write the `fiat_parsers` overlay without some help, so we can't expect an easy solution to work in, say, Coq 8.4.

**Comment by JasonGross** [comment]
> Or maybe, a quick and dirty solution would be to integrate a copy of Int31 directly in `fiat`?

This would be ideal: run https://github.com/JasonGross/coq-scripts/blob/master/compatibility/bump-versions.sh, replace https://github.com/JasonGross/coq-scripts/blob/8b66ebec09d641e5726d9931d4c99ae538f71283/compatibility/generate-compat-files.sh#L37 with
```bash
GRAB_88v816_VERSIONS="8_16 8_17 8_18"
GRAB_88v819_VERSIONS="8_19 master"
```
copy https://github.com/JasonGross/coq-scripts/blob/master/compatibility/fragments/Grab88v816.v to Grab88v819.v, and remove `Require Coq.Numbers.Cyclic.Int31.Int31.`, adding `Module Export Numbers. Module Export Cyclic. Module Export Int31. Module Int31. <Contents of Int31.v> ...` to https://github.com/JasonGross/coq-scripts/blob/8b66ebec09d641e5726d9931d4c99ae538f71283/compatibility/fragments/Grab88v816.v#L37 (except in the 819 file), and then rerun generate-compat-files.sh

If this is too complicated, let me know, and I can do it myself.

**Comment by JasonGross** [comment]
Alternatively, I think Int31 is in fact unused in Fiat Parsers (and everything else that uses coq-scripts, though I'm not 100% sure), so you could also just try removing the offending lines from compatibility/fragments/Grab88v816.v and running generate-compat-files.sh.  (The other files are only used in earlier versions of Coq)

**Comment by Villetaneuse** [comment]
Thank you very much. I will give it a try tomorrow, and if it proves too difficult I'll ask for help.

**Comment by Villetaneuse** [comment]
I tried your second solution. It was very promising at first, (many things compiled, a lot more than before the change) but then the CI entered a loop with
```text
## Remaking _CoqProject for Coq v8.19
```

**Comment by JasonGross** [comment]
> but then the CI entered in a seemingly infinite loop with

CI, or local build?  If you're building locally, I think you can fix this problem by making the change in etc/coq-scripts, and running `git commit` *from inside etc/coq-scripts*, and then running `git commit` again from the Fiat directory. Fiat has some questionable logic that tries to prevent people from accidentally using outdated versions of coq-scripts, which apparently has this infinite loop side effect if you have deliberate uncommitted changes to coq-scripts.  (And I think if you don't run git commit in Fiat before running make, it'll go check out the previous version of coq-scripts and ignore your changes.). Sorry for the hassle.

> For your second solution, do I need to actually fork _your coq-scripts repo_ instead of `fiat`?

Yes.  And the PR needs to be submitted to coq-scripts, and our automation will update Fiat within a day or so.

> And then how do I proceed to test the changes?

You can point the submodule at the new coq-scripts commit that you pushed, and make an overlay with that commit.  But in this case, if you make a PR to coq-scripts, I can just spot check it and merge (it should only be like 2--3 lines of change, right?), and then dependabot will open a PR on Fiat and we can see if Fiat's CI is happy with the change, and if not, fix it up after the fact.

**Comment by proux01** [comment]
@Villetaneuse what's the status of this?

**Comment by Villetaneuse** [comment]
The PRs are accepted in coqprime and metacoq. This should be over in a few days after the compat files are completed in Jason Gross' repo.

**Comment by JasonGross** [comment]
Sorry for the delay on my end, I had missed the PR for a while.  (Note that currently the PR is waiting on an update at https://github.com/JasonGross/coq-scripts/pull/11#issuecomment-1741932226 )

**Comment by Villetaneuse** [comment]
> Sorry for the delay on my end, I had missed the PR for a while. (Note that currently the PR is waiting on an update at [JasonGross/coq-scripts#11 (comment)](https://github.com/JasonGross/coq-scripts/pull/11#issuecomment-1741932226) )

I hope it is ok now. Sorry everyone for the delay. If it's ok with Jason Gross, it will move on.

**Comment by JasonGross** [comment]
Fiat should be good to go now with https://github.com/mit-plv/fiat/pull/101, @coqbot run full ci

**Comment by Villetaneuse** [comment]
I have just rebased on master and removed the overlays (since the PRs have been merged in metacoq and coqprime).

**Comment by Villetaneuse** [comment]
@coqbot run full ci

**Comment by proux01** [comment]
@coqbot run full ci

**Comment by Villetaneuse** [comment]
@coqbot run full ci

**Comment by Zimmi48** [comment]
The bug with coqbot should hopefully be fixed when https://github.com/Zimmi48/bot/actions/runs/6455747932 finishes (in about 5 minutes). Please retry running the full CI then.

**Comment by Zimmi48** [comment]
@coqbot run full CI

**Comment by proux01** [comment]
@coqbot merge now

**Comment by proux01** [comment]
@Villetaneuse thanks a lot for your patience with this!

**Comment by Villetaneuse** [comment]
> @Villetaneuse thanks a lot for your patience with this!

Thank you everyone for your patience, kindness and guidance! Especially @proux01 and @JasonGross.
